related-files.el

```diff
@@ -277,44 +333,76 @@ Beyond the filler, this function is called with the :jumper and
 
 ;;; Functions Manipulating Places
 
-(defun related-files--choose-place (places initial-place)
+(defun related-files--get-current-places ()
+  "Return a list of different forms of the current place."
+  (delete-dups
+   (seq-remove
+    #'null
+    (mapcar #'funcall related-files-current-place-finders))))
+
+(defun related-files--choose-place (places initial-places)
   "Let the user pick one of PLACES and return it.
 
-PLACES is a list of filenames and INITIAL-PLACE is a filename.
+PLACES is a list of filenames and INITIAL-PLACES is a list of
+places.
 
 INITIAL-PLACE is the place that was current when the user started
 related-files.  It is used to format each place in PLACES."
   (cond
    ((length= places 0) (user-error "No place to go to.  Consider configuring `related-files-jumpers' or using `related-files-make'") nil)
    ((length= places 1) (car places))
-   (t (let ((initial-directory (file-name-directory initial-place)))
-        (related-files--completing-read "Place: " places (apply-partially #'related-files--format-place initial-directory))))))
+   (t (related-files--completing-read "Place: " places (apply-partially #'related-files--format-place initial-places)))))
 
 (defun related-files--act-on-place (place)
   "Either open or create PLACE, a filename."
-  (if (file-exists-p place)
-      (find-file place)
+  (if (related-files-place-exists-p place)
+      (related-files-goto-place place)
     (related-files--make-place place)))
 
-(defun related-files--format-place (initial-directory place)
+(defcustom related-files-place-category-tests
+  '(((pred stringp) 'file)
+    ((pred bufferp) 'buffer))
+  "List of tests for getting type of a place.
+
+This variable is passed as an argument list to `pcase', so must
+take the relevant form. The return value of each clause must be a
+symbol for a category. See info node `(elisp)Programmed
+Completion' for more information."
+  ;; TODO pcase arg list type? (for :type)
+  :group 'related-files)
+
+(defmacro related-files--get-place-category (place)
+  "Get completion category for PLACE.
+
+Use `pcase' to check PLACE against the patterns in
+`related-files-place-category-tests'. If none succeed, return
+`type-of' PLACE."
+  `(pcase ,place
+     ,@related-files-place-category-tests
+     (t (type-of ,place))))
+
+(defun related-files--format-place (initial-places place &optional annotate)
```

Comment:
instead of having the `annotate` boolean, I would probably split this function into 2: one that never add "(create it!)" and one that does when the place doesn't exist.

Reply:
I introduced the optional `annotate` arg so that a single formatting function could be passed to `related-files--completing-read`, which would handle both annotated and non-annotated output -- this is important for compatibility with embark (see the `format-function` let-var in `related-files--completing-read`). Separating the function in two would mean this no longer worked, and would make the code inside that function much more complex. Personally I don't think that's worth it.

Reply:
very good! Thanks you for explaining.